### PR TITLE
lxd: Add VMs support

### PR DESCRIPTION
**Pull request by ifd3f**
###### Description of changes

This PR updates the changes made in #105651 by about 12 months. 

Original description:
> LXD has support for virtual machines, but this support isn't present in nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [x] NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### More info about testing

I ran the NixOS tests and they passed successfully.

This PR does not currently include automated tests for VMs, but I did perform manual tests. After building, I ran `sudo result/bin/lxd`, then used the `lxc` CLI to successfully launch a Fedora VM. I was additionally able to interact with `lxd-agent` and run bash inside the VM.

<details>
  <summary>Shell output</summary>

```
$ sudo lxc launch images:fedora/35/cloud testvm -s data -n br0 --vm
Creating testvm
Starting testvm

$ sudo lxc exec testvm bash
[root@testvm ~]#

$ sudo lxc info testvm
Name: testvm
Status: RUNNING
Type: virtual-machine
Architecture: x86_64
PID: 2467145
Created: 2022/04/01 11:57 PDT
Last Used: 2022/04/01 11:57 PDT

Resources:
  Processes: 18
  Disk usage:
    root: 11.55MiB
  CPU usage:
    CPU usage (in seconds): 0
  Network usage:
    enp5s0:
      Type: broadcast
      State: UP
      Host interface: tap9048ab67
      MAC address: 00:16:3e:42:a2:25
      MTU: 1500
      Bytes received: 2.27kB
      Bytes sent: 3.12kB
      Packets received: 14
      Packets sent: 25
      IP addresses:
        inet:  192.168.1.219/24 (global)
        inet6: fd73:e7aa:784e::20e/128 (global)
        inet6: fd73:e7aa:784e:0:216:3eff:fe42:a225/64 (global)
        inet6: fe80::216:3eff:fe42:a225/64 (link)
    lo:
      Type: loopback
      State: UP
      MTU: 65536
      Bytes received: 0B
      Bytes sent: 0B
      Packets received: 0
      Packets sent: 0
      IP addresses:
        inet:  127.0.0.1/8 (local)
        inet6: ::1/128 (local)
```

</details>

EDIT: I have verified that it works seamlessly on NixOS if you override the package as follows.

```nix
{
  virtualisation = {
    lxd = {
      enable = true;
      package = pkgs.lxd.override { useQemu = true; };
      recommendedSysctlSettings = true;
    };
  };
  boot.kernelModules = [ "vhost_vsock" ];
}
```

### CODE REVIEW

File: pkgs/tools/admin/lxd/default.nix [hunk 1]
```diff
@@ -1,64 +1,70 @@
...
 
-  postInstall = ''
-    # test binaries, code generation
-    rm $out/bin/{deps,macaroon-identity,generate}
+  binPath = lib.makeBinPath (
+    [ iptables ]
+    ++ [ acl rsync gnutar xz btrfs-progs gzip dnsmasq squashfsTools iproute2 bash criu attr ]
+    ++ lib.optionals useQemu [ qemu-utils qemu_kvm lxdAgent ]
+    ++ [ apparmor_parser ]
+    ++ [ "$out" ]
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:39]:
```suggestion
    [ apparmor_parser iptables acl rsync gnutar xz btrfs-progs gzip dnsmasq squashfsTools iproute2 bash criu attr ]
    ++ lib.optionals useQemu [ qemu-utils qemu_kvm lxdAgent ]
    ++ [ "$out" ]
```

File: pkgs/tools/admin/lxd/default.nix [hunk 2]
```diff
@@ -1,64 +1,70 @@
...
+in
+symlinkJoin {
+  name = "${pname}-${version}";
 
-    installShellCompletion --bash --name lxd go/src/github.com/lxc/lxd/scripts/bash/lxd-client
-  '';
+  paths = [ lxdBin ];
 
-  passthru.tests.lxd = nixosTests.lxd;
+  postBuild = "wrapProgram $out/bin/lxd --prefix PATH : ${binPath}"
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:48]:
```suggestion
  postInstall = "wrapProgram $out/bin/lxd --prefix PATH : ${binPath}"
```

File: pkgs/tools/admin/lxd/default.nix [hunk 3]
```diff
@@ -1,64 +1,70 @@
...
+    + lib.optionalString useQemu " --set LXD_OVMF_PATH $out/share/OVMF"
+    + lib.optionalString useQemu ''
+
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:51]:
```suggestion
    + lib.optionalString useQemu ''--set LXD_OVMF_PATH $out/share/OVMF
```

File: pkgs/tools/admin/lxd/lxd-agent.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, buildGoPackage
+, src, version
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd-agent.nix:3]:
```suggestion
, src
, version
```

File: pkgs/tools/admin/lxd/lxd-agent.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+}:
+
+buildGoPackage rec {
+  inherit version src;
+
+  pname = "lxd-agent";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd-agent.nix:9]:
```suggestion
  pname = "lxd-agent";
  inherit version src;
```

File: pkgs/tools/admin/lxd/lxd.nix [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+, criu
+, bash
+, installShellFiles
+, version, src
+}:
+
+buildGoPackage rec {
+  inherit version src;
+
+  pname = "lxd-unwrapped";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd.nix:15]:
```suggestion
  pname = "lxd-unwrapped";
  inherit version src;
```

File: pkgs/tools/admin/lxd/lxd.nix [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  postPatch = ''
+    substituteInPlace shared/usbid/load.go \
+      --replace "/usr/share/misc/usb.ids" "${hwdata}/share/hwdata/usb.ids"
+  '';
+
+  preBuild = ''
+    # required for go-dqlite. See: https://github.com/lxc/lxd/pull/8939
+    export CGO_LDFLAGS_ALLOW="(-Wl,-wrap,pthread_create)|(-Wl,-z,now)"
+
+    makeFlagsArray+=("-tags libsqlite3")
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd.nix:28]: Please use tags attribute instead.

File: pkgs/tools/admin/lxd/lxd.nix [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  '';
+
+  postInstall = ''
+    installShellCompletion --bash --name lxd go/src/github.com/lxc/lxd/scripts/bash/lxd-client
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd.nix:32]:
```suggestion
    installShellCompletion --cmd lxd --bash go/src/github.com/lxc/lxd/scripts/bash/lxd-client
```

File: pkgs/tools/admin/lxd/lxd.nix [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  '';
+
+  nativeBuildInputs = [ installShellFiles pkg-config ];
+  buildInputs = [ lxc acl libcap dqlite.dev raft-canonical.dev
+                  sqlite-replication udev.dev ];
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd.nix:37]:
```suggestion
  nativeBuildInputs = [ installShellFiles pkg-config ];

  buildInputs = [
    lxc
    acl
    libcap
    dqlite.dev
    raft-canonical.dev
    sqlite-replication
    udev.dev
  ];
```

Did you check if we still need `.dev`?

**Comment by astralbijection** [comment]
Currently, there seems to be an issue with nixostests:
```
machine # [   11.315508] lxd[909]: Error: exec: "ip": executable file not found in $PATH
machine # [   11.326115] systemd[1]: lxd.service: Main process exited, code=exited, status=1/FAILURE
```
I'll be looking into this issue.

**Comment by astralbijection** [comment]
I've fixed that issue, so the NixOS tests pass. Also, I've confirmed that the `.dev` libraries aren't needed. @SuperSandro2000 could you look it over again?

File: pkgs/tools/admin/lxd/lxd.nix [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    libcap
+    dqlite.dev
+    raft-canonical.dev
+    sqlite-replication
+    udev.dev
+  ];
+
+  meta = with lib; {
+    description =
+      "Daemon based on liblxc offering a REST API to manage containers";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd.nix:66]:
```suggestion
    description = "Daemon based on liblxc offering a REST API to manage containers";
```

File: pkgs/tools/admin/lxd/lxd.nix [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    homepage = "https://linuxcontainers.org/lxd/";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ fpletz marsam astralbijection ];
+    platforms = platforms.linux;
+  };
+
+  passthru.tests.lxd = nixosTests.lxd;
+}
+
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/lxd.nix:75]: passthru should be above meta, as meta should always be last.

**Comment by astralbijection** [comment]
I'm applying patches to this PR to use 5.0.0.

**Comment by astralbijection** [comment]
I added another NixOS test for when we set `useQemu = true` on the package. It's the same as the regular one for now, but we can add VM tests into it if we eventually upload certain files to the server.

**Comment by kellianoy** [comment]
Hello, I am using your pull request to launch VMs using LXD in NixOS. However, I can't seem to be able to make it work. My error is something along the lines of:

**What I do:** 

```
lxc launch ubuntu:16.04 --vm
```

**Error:**

```
Error: Failed instance creation: Failed creating instance from image: Failed setting nodatacow on "/var/lib/lxd/storage-pools/local/images/fdfe665ab7484ae2c6cbf043f66d112d4195a159c7331baf042330023e67531d": Failed to run: chattr +C /var/lib/lxd/storage-pools/local/images/fdfe665ab7484ae2c6cbf043f66d112d4195a159c7331baf042330023e67531d:
```

**Comment by astralbijection** [comment]
@kellianoy what storage backend are you using? Seems to work fine on ZFS for me. Also, when did you last pull from my fork?

**Comment by kellianoy** [comment]
@astralbijection I'm using BTRFS, as you mentionned. I pulled your branch 5 hours ago.

I don't actually have any data on my pool, so it is not even a problem. I'll try switching to ZFS, see if it resolves the problem.

Thank you for the quick answer !

**Comment by kellianoy** [comment]
@astralbijection Have you set specific parameters to enable ZFS on your fork branch ?

**Comment by kellianoy** [comment]
@astralbijection I can see that you tried to implement what [stevenewey](https://github.com/stevenewey/lxd-vms-on-nixos) did in his `lxd-vms-on-nixos` repo, but it doesn't seem to work, as I still do have lxd-agent problems. 
I will try to fork your repo and see if I can come up with a different solution, because I need both `lxd-agent` and `virtfs-proxy-helper`.
I don't think you implemented his [virtiofsd.nix](https://github.com/stevenewey/lxd-vms-on-nixos/blob/master/etc/nixos/virtiofsd.nix) file. Have you encountered limitations or is it simply that you didn't need this functionnality ?
Best regards.

File: maintainers/maintainer-list.nix [hunk 12]
```diff
@@ -1048,6 +1048,12 @@
     githubId = 453170;
     name = "Alastair Pharo";
   };
+  astralbijection = {
```
Comment (raboof) [inline maintainers/maintainer-list.nix:1051]: Welcome!

It's custom to have the addition to the `maintainer-list.nix` in a separate commit (https://nixos.org/manual/nixpkgs/unstable/#var-meta-maintainers).

Apologies for only commenting on such a trivial thing and not on the 'meat' of this PR, I'm afraid I don't feel qualified to judge the other changes...

**Comment by jmgilman** [comment]
Hi @ifd3f , would you mind addressing the latest comment so we can get this merged? Thanks!

**Comment by ifd3f** [comment]
@jmgilman will do. Sorry, I got a bit distracted from this due to school, but I'll figure out how to update and rebase this commit.

**Comment by adamcstephens** [comment]
Should VM support be enabled by default? Upstream treats VMs as first class guests. New users may expect VMs to be supported with the default package.

I think this is worth emerging either way, and am happy to see it added.

**Comment by ifd3f** [comment]
Sure, we could do that.

File: maintainers/maintainer-list.nix [hunk 13]
```diff
@@ -15893,4 +15893,10 @@
     github = "wuyoli";
     githubId = 104238274;
   };
+  ifd3f = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:15896]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/tools/admin/lxd/default.nix [hunk 14]
```diff
@@ -70,41 +83,42 @@ buildGoModule rec {
...
+  passthru.tests = {
+    lxd = nixosTests.lxd;
+    lxd-nftables = nixosTests.lxd-nftables;
+    lxd-qemu = nixosTests.lxd-qemu;
+  };
 
   meta = with lib; {
-    description = "Daemon based on liblxc offering a REST API to manage containers";
+    description =
+      "Daemon based on liblxc offering a REST API to manage containers";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:117]:
```suggestion
    description = "Daemon based on liblxc offering a REST API to manage containers";
```

File: pkgs/tools/admin/lxd/default.nix [hunk 15]
```diff
@@ -70,41 +83,42 @@ buildGoModule rec {
...
 
     installShellCompletion --bash --name lxd ./scripts/bash/lxd-client
   '';
 
-  passthru.tests.lxd = nixosTests.lxd;
-  passthru.tests.lxd-nftables = nixosTests.lxd-nftables;
+  passthru.tests = {
+    lxd = nixosTests.lxd;
+    lxd-nftables = nixosTests.lxd-nftables;
+    lxd-qemu = nixosTests.lxd-qemu;
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:112]:
```suggestion
    inherit (nixosTests) lxd lxd-nftables lxd-qemu;
```

**Comment by ifd3f** [comment]
@SuperSandro2000 I applied the suggested changes, could you look over it again?

**Comment by adamcstephens** [comment]
I've been unable to get this PR working with lxd 5.8:

```
─❯ lxc launch images:alpine/3.17 t1 --vm -c security.secureboot=false
Error: Failed parsing vsock address: input does not match format
```

This section of code was modified https://github.com/lxc/lxd/pull/11080 which was in the 5.8 release.

I tried rolling the PR back to lxd 5.7. Which allows me to create a VM, but they crash on start. 

```
KVM internal error. Suberror: 1
extra data[0]: 0x0000000000000000
extra data[1]: 0x0000000000000400
extra data[2]: 0x0000000100000014
extra data[3]: 0x00000000000a0000
extra data[4]: 0x0000000000000000
extra data[5]: 0x0000000000000000
emulation failure
EAX=0000ef11 EBX=00000272 ECX=00000271 EDX=00000103
ESI=00000000 EDI=00000271 EBP=0000000c ESP=0000aa8c
EIP=0008b000 EFL=00003002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =6fe8 0006fe80 ffffffff 00809300
CS =1500 00015000 ffffffff 00809b00
SS =5000 00050000 ffffffff 00809300
DS =0000 00000000 ffffffff 00809300
FS =0000 00000000 ffffffff 00809300
GS =0000 00000000 ffffffff 00809300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     000f6720 00000037
IDT=     00000000 000003ff
CR0=00010032 CR2=00000000 CR3=3fc01000 CR4=00000648
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000800
Code=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 <00> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

File: pkgs/tools/admin/lxd/default.nix [hunk 16]
```diff
@@ -42,6 +14,47 @@ buildGoModule rec {
...
+    btrfs-progs
+    criu
+    dnsmasq
+    gnutar
+    gzip
+    iproute2
+    iptables
+    rsync
+    squashfsTools
+    xz
```
Comment (adamcstephens) [inline pkgs/tools/admin/lxd/default.nix:30]: I suggest adding some more packages to this list

- `util-linux` I needed this to even create a VM instance, otherwise it errors trying to for `prlimit`
- `gptfdisk` VM extraction tries to use `sfdisk`
- `nftables` for firewall support when nftables is enabled

**Review by adamcstephens — CHANGES_REQUESTED** [review]
I've been unable to get this PR working with lxd 5.8:

```
─❯ lxc launch images:alpine/3.17 t1 --vm -c security.secureboot=false
Error: Failed parsing vsock address: input does not match format
```

This section of code was modified https://github.com/lxc/lxd/pull/11080 which was in the 5.8 release. I spent some time troubleshooting this but need to dig deeper.

I tried rolling the PR back to lxd 5.7. Which allows me to create a VM, but they crash on start. This was also run with `images:alpine/3.17`.

```
KVM internal error. Suberror: 1
extra data[0]: 0x0000000000000000
extra data[1]: 0x0000000000000400
extra data[2]: 0x0000000100000014
extra data[3]: 0x00000000000a0000
extra data[4]: 0x0000000000000000
extra data[5]: 0x0000000000000000
emulation failure
EAX=0000ef11 EBX=00000272 ECX=00000271 EDX=00000103
ESI=00000000 EDI=00000271 EBP=0000000c ESP=0000aa8c
EIP=0008b000 EFL=00003002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =6fe8 0006fe80 ffffffff 00809300
CS =1500 00015000 ffffffff 00809b00
SS =5000 00050000 ffffffff 00809300
DS =0000 00000000 ffffffff 00809300
FS =0000 00000000 ffffffff 00809300
GS =0000 00000000 ffffffff 00809300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     000f6720 00000037
IDT=     00000000 000003ff
CR0=00010032 CR2=00000000 CR3=3fc01000 CR4=00000648
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000800
Code=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 <00> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

File: nixos/tests/lxd-qemu.nix [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+      # Our tests require `grep` & friends:
+      environment.systemPackages = with pkgs; [ busybox ];
+    };
+  };
+
+  lxd-image-metadata = lxd-image.lxdMeta.${pkgs.stdenv.hostPlatform.system};
+  lxd-image-rootfs = lxd-image.lxdImage.${pkgs.stdenv.hostPlatform.system};
+
+in {
+  name = "lxd";
```
Comment (marsam) [inline nixos/tests/lxd-qemu.nix:23]:
```suggestion
  name = "lxd-qemu";
```

File: pkgs/tools/admin/lxd/default.nix [hunk 18]
```diff
@@ -1,39 +1,11 @@
...
-
-buildGoModule rec {
-  pname = "lxd";
+{ lib, hwdata, pkg-config, lxc, buildGoModule, fetchurl, makeWrapper, acl, rsync
+, gnutar, xz, btrfs-progs, gzip, dnsmasq, attr, squashfsTools, iproute2
+, iptables, libcap, dqlite, raft-canonical, sqlite-replication, udev
+, writeShellScriptBin, apparmor-profiles, apparmor-parser, criu, bash
+, callPackage, installShellFiles, linkFarm, qemu_kvm, qemu-utils, OVMFFull
+, nixosTests, useQemu ? true }:
+let
```
Comment (marsam) [inline pkgs/tools/admin/lxd/default.nix:7]: please revert the formatting changes

**Comment by FlyingWombat** [comment]
LXD needs modprobe from kmod package and mkisofs from cdrkit package in binPath, when working with VMs.

**Comment by adamcstephens** [comment]
> I tried rolling the PR back to lxd 5.7. Which allows me to create a VM, but they crash on start. This was also run with `images:alpine/3.17`.
> 
> ```
> KVM internal error. Suberror: 1
> ...
> ```

After much troubleshooting, I discovered this is due to the OVMF files. I was able to copy a set of OVMF files from the LXD snap and start VMs. I tried bumping edk2 (and the downstream OVMF) in #204052 but that made no difference. So I suspect maybe they're being generated incorrectly? I verified this on both an AMD nested VM and an Intel physical machine.

I could replicate the failure using the following command. With the nixpkgs OVMF code I get no serial output. With the LXD snap OVMF code it outputs its attempts to boot from the (missing) disk and network.

```
/nix/store/n12432cd9biha7wgx0q9pa33ysyxd296-qemu-host-cpu-only-7.1.0/bin/qemu-system-x86_64 -drive if=pflash,format=raw,unit=0,readonly=on,file=OVMF_CODE.fd -nographic -enable-kvm
```

@ifd3f does this PR work for you? Are you able to successfully create VMs with it?

**Comment by AmandaCameron** [comment]
@adamcstephens It seems to be something with OVMFull being used, which was introduced in oct. I think. I never got around to debugging why, but I managed to fix my own little hack around this in my dotfiles by using OVMF instead of OVMFFull, but with your PR to bump the version, now it seems to just be booting into an UEFI shell on my VMs. I think it might have somethign to do with the csmSupport option which is causing the KVM error.

**Comment by AmandaCameron** [comment]
Sorry, I kinda word-salad'd that message.

To clarify: the KVM error appears to be caused by the OVMF option in the nixpkgs for "csmSupport" -- I previously worked around the KVM option in a different bit of lxd fuckery I have for my systems by using OVMF instead of OVMFFull from nixpkgs, however your recent PR to bump OVMF made it so that it seems to always be booting into an UEFI shell.

**Comment by adamcstephens** [comment]
> using OVMF instead of OVMFFull

Switching to OVMF works for me as well.

File: pkgs/tools/admin/lxd/default.nix [hunk 19]
```diff
@@ -42,6 +14,47 @@ buildGoModule rec {
...
+    iproute2
+    iptables
+    rsync
+    squashfsTools
+    xz
+
+    (writeShellScriptBin "apparmor_parser" ''
+      exec '${apparmor-parser}/bin/apparmor_parser' -I '${apparmor-profiles}/etc/apparmor.d' "$@"
+    '')
+  ] ++ (lib.optionals useQemu [ qemu-utils qemu_kvm ]);
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:35]:
```suggestion
  ] ++ lib.optionals useQemu [ qemu-utils qemu_kvm ];
```

File: pkgs/tools/admin/lxd/default.nix [hunk 20]
```diff
@@ -1,39 +1,11 @@
...
-  pname = "lxd";
+{ lib, hwdata, pkg-config, lxc, buildGoModule, fetchurl, makeWrapper, acl, rsync
+, gnutar, xz, btrfs-progs, gzip, dnsmasq, attr, squashfsTools, iproute2
+, iptables, libcap, dqlite, raft-canonical, sqlite-replication, udev
+, writeShellScriptBin, apparmor-profiles, apparmor-parser, criu, bash
+, callPackage, installShellFiles, linkFarm, qemu_kvm, qemu-utils, OVMFFull
+, nixosTests, useQemu ? true }:
+let
   version = "5.8";
-
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/default.nix:36]: Keeping some new lines between things increases readability

File: nixos/tests/lxd-qemu.nix [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+    # lxd expects the pool's directory to already exist
+    machine.succeed("mkdir /var/lxd-pool")
+
+    machine.succeed(
+        "cat ${./common/lxd/config.yaml} | lxd init --preseed"
+    )
+
+    machine.succeed(
+        "lxc image import ${lxd-image-metadata}/*/*.tar.xz ${lxd-image-rootfs}/*/*.tar.xz --alias nixos"
```
Comment (mkg20001) [inline nixos/tests/lxd-qemu.nix:61]: this is currently running the test using the regular (cgroup) container type

for this to be fixed there would need to be a qemu lxd image. this would be mostly a basic qemu image, plus the lxd defaults and wrapped in tar with metadata

**Review by mkg20001 — CHANGES_REQUESTED** [review]

**Comment by InLaw** [comment]
is there any open topic besides the code formatting?

**Comment by adamcstephens** [comment]
The wrapper changes are already in, and I've staged a few more PRs towards this effort. Thank you for your effort, it was incredibly helpful for me as I worked through this.

I'm going to close, and recommend watching #244093 and the attached PRs.
